### ISSUE TITLE
Dozeu's allocator for long reads

### ISSUE THREAD

**Issue by StephenHwang** [vgteam/vg#3733]
**1. What were you trying to do?**

Use VG Giraffe to align long giraffe reads.

**2. What did you want to happen?**

Successfully align long giraffe reads.

**3. What actually happened?**

Segmentation fault and warning from vg. No usual explicit `stacktrace + make issue`  error from vg.

**4. If you got a line like `Stack trace path: /somewhere/on/your/computer/stacktrace.txt`, please copy-paste the contents of that file here:**

No explicit given stacktrace; program broke on `Segmentation fault`

Warning was given directly before:
```
warning[vg::giraffe]: Refusing to perform too-large tail alignment of 15767 bp against 1159639 bp tree which would use more than 1572864 cells and might exhaust Dozeu's allocator; suppressing further warnings.
```

On other datasets, the same warning is given but Giraffe continues running. One such example dataset is a simulated Nanopore dataset (simulated from the reads that caused this seg fault): `/public/groups/vg/sjhwang/vg_scripts/bin/reads/sim_HiFi_other_tools/sim_pbsim2/sim_NA19239/gam/annotated/concat/NA19239_100k.gam`

**5. What data and command can the vg dev team use to make the problem happen?**

The script to reproduce the error is located on UCSC courtyard: `/public/groups/vg/sjhwang/vg_scripts/bin/debug/run_giraffe_dozeu_allocator.sh`
The output log warning is also located in the same directory.

**6. What does running `vg version` say?**

```
vg version v1.42.0-112-g849d977c1 "Obolo"
Compiled with g++ (GCC) 11.2.0 on Linux
Linked against libstd++ 20210728
Built by sjhwang@courtyard
```

**Comment by StephenHwang** [comment]
@adamnovak Here's the issue on Dozeu's allocator

**Comment by adamnovak** [comment]
I've managed to reproduce this, I think.

I had to let it rebuild the minimizer index, because I was getting:

```
MinimizerIndex::deserialize(): Expected 64-bit keys, got 128-bit keys
```

So now I'm running

```
vg giraffe -p --track-provenance --track-correctness --exclude-overlapping-min --threads 32 -H /public/home/anovak/build/vg/trash/dozeu-crash/1000gp.gbwt -x /public/home/anovak/build/vg/trash/dozeu-crash/1000gp.xg -d /public/home/anovak/build/vg/trash/dozeu-crash/1000gp.dist -f /public/groups/vg/sjhwang/data/reads/real_ONT/HPRC_hg00423/HG00423_HPRC_yr2_1k_gap.fq --hit-cap 10 --hard-hit-cap 500 --score-fraction 1.0 --max-min 2000 >/dev/null
```

I managed to get it to crash with that after about half an hour. I re-ran under gdb and I got a segfault caused by what looks to be a stack overflow under `vg::MinimizerMapper::get_tail_forest`, for read `b8b0f580-4279-4364-ae3c-a29398065629`. I think we designed the Giraffe tail alignment to just recurse all the way to the end of the read in `vg::MinimizerMapper::dfs_gbwt`, and that doesn't work with the long reads anymore because that can involve visiting more nodes than we have stack space to recurse through.

This makes sense to me, since I was pretty sure we actually already walled off Dozeu's allocator from ever being asked to do anything too strenuous that could make it exhaust its internal space.

The fix for the segfault is to redesign our tail tree finding algorithm to not be recursive anymore, which might be a lot of work. And that still won't fix Dozeu not being willing to actually align the long tails, which would require actually implementing the missing pieces of its allocator and then removing our protective safeties.

@StephenHwang You are wanting this fixed so you can get your test to run through, right? Not because you actually want to be able to get good alignments quickly all the way out to the ends of the reads from the old short-read-Giraffe algorithm?

**Comment by jeizenga** [comment]
FWIW I'm also working on straightening out dozeu's memory allocation right now.

**Comment by adamnovak** [comment]
I think we decided this wasn't really the allocator, and fixed it in #3738.
